### DISCUSSION TITLE
`dump_bulk` alternative for `django-treenode`

### DISCUSSION THREAD

**Discussion by baseplate-admin** [fabiocaccamo/django-treenode#89] [Q&A] [answered]
Hi,

Thanks for creating `django-treenode`. I was thinking if it was possible to have the [`dump_bulk`](https://django-treebeard.readthedocs.io/en/latest/api.html?highlight=bulk_dump#treebeard.models.Node.dump_bulk) api or something like that in `django-treenode`

So essentially i can convert this model ( image ) :
<p align="center">

<img  src= 'https://user-images.githubusercontent.com/61817579/229573830-91945767-84f1-4384-923d-1a6af7c3d7c0.png' alt='graphviz-43a7e84ebb5c9b243f3b4beccb5489556e74fc18'/>

</p>

To 

```python
data = [
    {"data": {"desc": "1"}},
    {
        "data": {"desc": "2"},
        "children": [
            {"data": {"desc": "21"}},
            {"data": {"desc": "22"}},
            {
                "data": {"desc": "23"},
                "children": [
                    {"data": {"desc": "231"}},
                ],
            },
            {"data": {"desc": "24"}},
        ],
    },
    {"data": {"desc": "3"}},
    {
        "data": {"desc": "4"},
        "children": [
            {"data": {"desc": "41"}},
        ],
    },
]

```

**Comment by fabiocaccamo** [comment, marked as answer]
Hi @baseplate-admin 
you can try to use the model class method to obtain a *n-dimensional* dict representing the model tree:
```python
YourModelClass.get_tree()
```

Alternatively, if you need to have more control on the desired output, you could:

1. Use Django [`dumpdata`](https://docs.djangoproject.com/en/4.1/ref/django-admin/#django-admin-dumpdata) command to output a flat json of all nodes
2. Use [`python-benedict`](https://github.com/fabiocaccamo/python-benedict) for loading the `.json` and building the dict tree with two lines of code (using the [`nest`](https://github.com/fabiocaccamo/python-benedict#nest) method)

I hope to have been helpful.

**Reply by baseplate-admin** [reply]
Got it. I will have a look. Thanks again for input

**Reply by fabiocaccamo** [reply]
Let me know!

**Reply by baseplate-admin** [reply]
> Use [python-benedict](https://github.com/fabiocaccamo/python-benedict) for loading the .json and building the dict tree with two lines of code (using the [nest](https://github.com/fabiocaccamo/python-benedict#nest) method)

Would you mind adding this to the `README.md` ?

**Reply by fabiocaccamo** [reply]
> > Use [python-benedict](https://github.com/fabiocaccamo/python-benedict) for loading the .json and building the dict tree with two lines of code (using the [nest](https://github.com/fabiocaccamo/python-benedict#nest) method)
> 
> 
> 
> Would you mind adding this to the `README.md` ? 

That was your solution? Have you tried the `get_tree` static method?

**Reply by baseplate-admin** [reply]
> That was your solution? Have you tried the get_tree static method?

`get_descendants_tree` did also work for me.. But i needed a custom data structure. Thats why i followed the custom approach documented below. 

All in all thanks my man for helping me out here... You rock :D

**Reply by fabiocaccamo** [reply]
I added a link to this discussion in a new FAQ section of the README:
https://github.com/fabiocaccamo/django-treenode#custom-tree-serialization

**Comment by baseplate-admin** [comment]
Hi,

so i am making reddit style tree comment system for `django-ninja`.

The way i achieved this is with this code :

```python
# models.py
from apps.user.models import CustomUser

from django.db import models

from treenode.models import TreeNodeModel

# Create your models here.


class EpisodeCommentModel(TreeNodeModel):
    user = models.ForeignKey(CustomUser, on_delete=models.CASCADE)
    text = models.TextField()

    comment_added = models.DateTimeField(auto_now=True)

    def __str__(self) -> str:
        return f"{self.user} | {self.text}"

    class Meta(TreeNodeModel.Meta):
        verbose_name = "Episode Comment"
        verbose_name_plural = "Episode Comments"

```

```python
# views.py

@router.get(
    "/{int:anime_id}/episodes/{str:episode_number}/comments",
    response=list[EpisodeCommentTreeSchema],
)
def get_individual_anime_episode_comments(
    request: HttpRequest,
    anime_id: int,
    episode_number: str,
):
    query: list[EpisodeCommentModel] = get_list_or_404(
        get_object_or_404(
            AnimeModel,
            pk=anime_id,
        )
        .episodes.get(episode_number__in=[episode_number])
        .episode_comments.all()
    )

    return_list = []

    def get_nested_children(item: EpisodeCommentModel):
        __list__ = []
        __list__.append(
            {
                "user": str(item.user),
                "text": item.text,
                "comment_added": item.comment_added,
                "children": [get_nested_children(i)[0] for i in item.get_children()],
            }
        )
        return __list__

    for i in query:
        return_list.append(get_nested_children(i)[0])

    return return_list

```


```python
# schema.py
class EpisodeCommentTreeSchema(Schema):
    user: str
    text: str
    comment_added: datetime.datetime
    children: list["EpisodeCommentTreeSchema"] = []
```

Which returns

```typescript
[
  {
    "user": "baseplate-admin#0001",
    "text": "1",
    "comment_added": "2023-04-04T15:37:17.350Z",
    "children": [
      {
        "user": "baseplate-admin#0001",
        "text": "2",
        "comment_added": "2023-04-03T16:17:53.050Z",
        "children": [
          {
            "user": "baseplate-admin#0001",
            "text": "3",
            "comment_added": "2023-04-03T16:18:13.588Z",
            "children": []
          }
        ]
      },
      {
        "user": "baseplate-admin#0001",
        "text": "2",
        "comment_added": "2023-04-04T15:51:39.823Z",
        "children": []
      }
    ]
  },
  {
    "user": "baseplate-admin#0001",
    "text": "2",
    "comment_added": "2023-04-03T16:24:58.779Z",
    "children": []
  }
]
```

---
<sub>

[Source](https://github.com/baseplate-admin/CoreProject/blob/83f7148a19ef49cf55880797044d67c8e39827fa/backend/django_core/apps/api/views/anime/episode_comment.py#L19-L55)

</sub>
